Generic function fails to be created on second use of generic function

### Describe the bug

There is a bug with the generics where it will not produce the generic function for the second use of the generic function. In this case, func2[Struct1] is generated, but func2[Struct2] is not, causing a compilation error

### Reproduction Steps

The following code will produce the issue
```v
module main

struct ClassInfo {
	func fn () = unsafe { nil }
}

pub fn func2[T]() {
}

pub fn func1[T]() {
	ci := ClassInfo{
		func: func2[T]
	}
	ci.func()
}

struct Struct1 {}

struct Struct2 {}

pub fn main() {
	func1[Struct1]()
	func1[Struct2]()
}

```

### Expected Behavior

successful compile

### Current Behavior

Fails to compile with the following error
```
PS C:\Users\jcwea\Documents\git\vtest> v -cg run .    
C:/Users/jcwea/AppData/Local/Temp/v_0/vtest.01HPSP98VD487P3GG4CG3VMKQV.tmp.c:497: warning: WINVER redefined
C:/Users/jcwea/AppData/Local/Temp/v_0/vtest.01HPSP98VD487P3GG4CG3VMKQV.tmp.c:6880: warning: implicit declaration of function 'tcc_backtrace'
C:/Users/jcwea/AppData/Local/Temp/v_0/vtest.01HPSP98VD487P3GG4CG3VMKQV.tmp.c:13104: error: 'main__func2_T_main__Struct2' undeclared
builder error:
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.4.4 1d3147e

### Environment details (OS name and version, etc.)

```
V full version: V 0.4.4 afd74ad.1d3147e
OS: windows, Microsoft Windows 11 Pro v26058 64-bit
Processor: 24 cpus, 64bit, little endian, 

getwd: C:\Users\jcwea\Documents\git\vtest
vexe: C:\Users\jcwea\Documents\git\v\v.exe
vexe mtime: 2024-02-16 19:26:43

vroot: OK, value: C:\Users\jcwea\Documents\git\v
VMODULES: OK, value: C:\Users\jcwea\.vmodules
VTMP: OK, value: C:\Users\jcwea\AppData\Local\Temp\v_0

Git version: git version 2.40.0.windows.1
Git vroot status: weekly.2024.06-86-g1d3147e1
.git/config present: true

CC version: Error: 'cc' is not recognized as an internal or external command,
operable program or batch file.

thirdparty/tcc status: thirdparty-windows-amd64 b99a453d
```
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/checker/assign.v
+++ b/vlib/v/checker/assign.v
@@ -189,8 +189,8 @@ fn (mut c Checker) assign_stmt(mut node ast.AssignStmt) {
 		if is_decl || is_shared_re_assign {
 			// check generic struct init and return unwrap generic struct type
 			if mut right is ast.StructInit {
+				c.expr(mut right)
 				if right.typ.has_flag(.generic) {
-					c.expr(mut right)
 					right_type = right.typ
 				}
 			} else if mut right is ast.PrefixExpr {

--- a/vlib/v/tests/struct_field_init_with_generic_fn_test.v
+++ b/vlib/v/tests/struct_field_init_with_generic_fn_test.v
@@ -1,0 +1,36 @@
+struct ClassInfo {
+	func fn () int = unsafe { nil }
+}
+
+pub fn func2[T]() int {
+	return sizeof(T)
+}
+
+pub fn func1[T]() int {
+	ci := ClassInfo{
+		func: func2[T]
+	}
+	return ci.func()
+}
+
+struct Struct1 {}
+
+struct Struct2 {}
+
+fn test_struct_field_init_with_generic_fn() {
+	r1 := func1[i32]()
+	println(r1)
+	assert r1 == 4
+
+	r2 := func1[bool]()
+	println(r2)
+	assert r2 == 1
+
+	r3 := func1[Struct1]()
+	println(r3)
+	assert r3 == 1
+
+	r4 := func1[Struct2]()
+	println(r4)
+	assert r4 == 1
+}

--- a/vlib/v/tests/struct_field_init_with_generic_fn_test.v
+++ b/vlib/v/tests/struct_field_init_with_generic_fn_test.v
@@ -1,16 +1,15 @@
 struct ClassInfo {
-	func fn () int = unsafe { nil }
+	func fn () = unsafe { nil }
 }
 
-pub fn func2[T]() int {
-	return sizeof(T)
+pub fn func2[T]() {
 }
 
-pub fn func1[T]() int {
+pub fn func1[T]() {
 	ci := ClassInfo{
 		func: func2[T]
 	}
-	return ci.func()
+	ci.func()
 }
 
 struct Struct1 {}
@@ -18,19 +17,9 @@ struct Struct1 {}
 struct Struct2 {}
 
 fn test_struct_field_init_with_generic_fn() {
-	r1 := func1[i32]()
-	println(r1)
-	assert r1 == 4
-
-	r2 := func1[bool]()
-	println(r2)
-	assert r2 == 1
-
-	r3 := func1[Struct1]()
-	println(r3)
-	assert r3 == 1
-
-	r4 := func1[Struct2]()
-	println(r4)
-	assert r4 == 1
+	func1[i32]()
+	func1[bool]()
+	func1[Struct1]()
+	func1[Struct2]()
+	assert true
 }

